api/bucketsd/service.go

```diff
@@ -970,16 +1073,184 @@ func (s *Service) SetPath(ctx context.Context, req *pb.SetPathRequest) (res *pb.
 	}
 	dbToken, _ := thread.TokenFromContext(ctx)
 
-	destPath := cleanPath(req.Path)
-	bootCid, err := cid.Decode(req.Cid)
+	fromPth, err := parsePath(req.FromPath)
 	if err != nil {
-		return nil, fmt.Errorf("invalid remote cid: %v", err)
+		return nil, err
+	}
+
+	toPth, err := parsePath(req.ToPath)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: enable move of root directory
+	// Can't move root directory
+	if fromPth == "" {
+		return nil, fmt.Errorf("source is root directory")
+	}
+
+	// Paths are the same, return
+	if fromPth == toPth {
+		return &pb.MovePathResponse{}, nil
+	}
+
+	// Lock bucket for the rest of transformations
+	lck := s.Semaphores.Get(buckLock(req.Key))
+	lck.Acquire()
+	defer lck.Release()
+
+	buck, pth, err := s.getBucketPath(ctx, dbID, req.Key, fromPth, dbToken)
+	if err != nil {
+		return nil, fmt.Errorf("getting path: %v", err)
+	}
+
+	fromNodePath, err := inflateFilePath(buck, fromPth)
+	if err != nil {
+		return nil, err
+	}
+	fromItem, err := s.pathToItem(ctx, buck, fromNodePath, false)
+	if err != nil {
+		return nil, err
+	}
+
+	toNodePath, err := inflateFilePath(buck, toPth)
+	if err != nil {
+		return nil, err
+	}
+
+	toItem, err := s.pathToItem(ctx, buck, toNodePath, false)
+	if err == nil {
+		if fromItem.IsDir && !toItem.IsDir {
+			return nil, fmt.Errorf("Not a directory")
+		}
+		if toItem.IsDir {
+			// from => to becomes cleanDst:
+			// "c" => "b" becomes "b/c"
+			// "a.jpg" => "b" becomes "b/a.jpg"
+			toPth = gopath.Join(toPth, fromItem.Name)
+		}
+	}
+
+	// IPNS node
+	buckNode, err := s.getNodeAtPath(ctx, pth, buck.GetLinkEncryptionKey())
+	if err != nil {
+		return nil, fmt.Errorf("getting node: %v", err)
+	}
+
+	var dirPath path.Resolved
+	// Private buckets need to manage keys+encrypted nodes
+	if buck.IsPrivate() {
+		dirPath, err = s.copyNode(ctx, buck, buckNode, fromPth, toPth)
+		if err != nil {
+			return nil, fmt.Errorf("copying node: %v", err)
+		}
+	} else {
+		buckPath := path.New(buck.Path)
+		_, dirPath, err = s.setPathFromExistingPath(ctx, buck, buckPath, fromPth, toPth, buckNode.Cid(), nil, nil)
+		if err != nil {
+			return nil, fmt.Errorf("error copying path: %v", err)
+		}
+	}
+
+	buck.Path = dirPath.String()
+	buck.UpdatedAt = time.Now().UnixNano()
+	buck.SetMetadataAtPath(toPth, tdb.Metadata{
+		UpdatedAt: buck.UpdatedAt,
+	})
+	buck.UnsetMetadataWithPrefix(fromPth)
+
+	if err = s.Buckets.Verify(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
+		return nil, fmt.Errorf("verifying bucket update: %v", err)
+	}
+	if err = s.Buckets.Save(ctx, dbID, buck, tdb.WithToken(dbToken)); err != nil {
```

Comment:
This should be done last after all the pins have successfully been updated.

Reply:
right... i do it here because all the remaining methods are still using the off the shelf/existing functions (e.g. removePath, ListPath) that will pull the buck object again. i'll swap those out now to use the already open buck